application/dash_application/views/bcl2barcode.py

```diff
@@ -57,33 +50,84 @@ def barcode_test_test(row):
 # Where there is no 10X barcode, fill in with the iusTag
 pinery_with_expanded_barcodes['Sequence'] = pinery_with_expanded_barcodes['Sequence'].fillna(pinery_with_expanded_barcodes['iusTag'])
 
-# Merge the expanded pinery with bcl2barcode to expand the bcl2barcode data the same way
-bcl2barcode_with_pinery = pandas.merge(bcl2barcode, pinery_with_expanded_barcodes, left_on=['Run Alias', 'Lane Number'], right_on=['sequencerRunName', 'laneNumber'], how='left', indicator=True)
-pdb.set_trace()
-bcl2barcode_with_pinery['Match'] = bcl2barcode_with_pinery.apply(lambda r: barcode_test_test(r), axis='columns')
+# Split up Index 1 and 2 into their own columns
+pinery_with_expanded_barcodes['Index1'] = pinery_with_expanded_barcodes['Sequence'].fillna(pinery_with_expanded_barcodes['iusTag'])
+pinery_with_expanded_barcodes['Index2'] = pinery_with_expanded_barcodes['Index1'].apply(lambda s: numpy.nan if len(s.split("-")) == 1 else s.split("-")[1])
+pinery_with_expanded_barcodes['Index1'] = pinery_with_expanded_barcodes['Index1'].apply(lambda s: s.split("-")[0])
+pinery_with_expanded_barcodes.loc[pinery_with_expanded_barcodes['Index1'] == 'NoIndex', 'Index1'] = numpy.nan
+
+# Remove Pinery records that don't have Index 1 or Index 2
+pinery_with_expanded_barcodes = pinery_with_expanded_barcodes[~(pinery_with_expanded_barcodes['Index1'].isna() & pinery_with_expanded_barcodes['Index1'].isna())]
+
+# Calculate mininum index length per run/lane
+min_index1_length = pinery_with_expanded_barcodes.groupby(['sequencerRunName', 'laneNumber'])['Index1'].apply(lambda x: x.str.len().min()).reset_index().rename(columns={'Index1': 'Index1 min length'})
```

Comment:
This has to be per sample. The lab has routine run things that are like:

CCCC
GGGGAT
GGGGCG

so truncating to 4 is not acceptable.

Reply:
We are in trade-off land and I am looking for consensus in approach.

In the per sample approach (b9e969e), it is necessary to merge `pinery` and `bcl2barcode` on lane/run, which already results in a huge explosion of rows (2 million) and will get big fast as `pinery` and `bcl2barcode` grows. Then in the huge DataFrame check if single/dual index, trim, and compare barcode for each row (and pandas is gonna be really slow at that as trimming cannot be vectorized).

My approach has the weakness you describe, but I think it can be fixed. The `known` DataFrame can be inspected to check for duplicated barcode/lane/run (`sum(known_data_table.duplicated(['Barcodes', 'Lane Number', 'Run Alias']))`. Currently, there are 0 such duplicates, which makes me a little suspicious of this approach. In future, the first step is to add a warning to Dashi view stating that run has possible index collision and then be even better and smartly check if it is an actual index duplication or a false positive due to my trimming approach.